src/features/login/pages/LoginPage/index.tsx

```diff
@@ -15,50 +14,110 @@ import {
   ErrorMessage,
 } from './styled'
 
+const validators = {
+  email: (value: string) => {
+    if (typeof value !== 'string') return 'Invalid e-mail value type'
+    if (!value) return 'E-mail is required'
+    if (!/^[^\s@]+@[^\s@]+\.[^\s@]+$/u.test(value)) return 'Invalid e-mail'
+  },
+
+  password: (value: string) => {
+    if (typeof value !== 'string') return 'Invalid password value type'
+    if (!value) return 'Password is required'
+  },
+}
+
 export const LoginPage: NextPage = () => {
-  const [error, setError] = useState('')
+  const [email, setEmail] = useState('')
+  const [password, setPassword] = useState('')
+
+  const [emailError, setEmailError] = useState<string | null>(null)
+  const [passwordError, setPasswordError] = useState<string | null>(null)
+  const [submitError, setSubmitError] = useState<string | null>(null)
+
+  const [isSubmitting, setIsSubmitting] = useState(false)
+
+  /**
+   * Login handler.
+   */
+  const login = useCallback(
+    (e: FormEvent<HTMLFormElement>) => {
+      e.preventDefault()
+
+      const errors = {
+        email: validators.email(email),
+        password: validators.password(password),
+      }
+
+      if (errors.email) {
+        setEmailError(errors.email)
+      }
+
+      if (errors.password) {
+        setPasswordError(errors.password)
+      }
+
+      // Only submit in case of no errors.
+      if (!errors.email && !errors.password) {
+        setIsSubmitting(true)
 
-  const onSubmit = (event: FormEvent<HTMLFormElement>) => {
-    event.preventDefault()
+        setTimeout(() => {
+          // Mocking to represent login submit outcome.
+          const shouldFail = Math.random() < 0.5
 
-    alert('TODO')
-  }
+          if (shouldFail) {
+            setSubmitError('Something went terribly wrong!')
+          } else {
+            alert('Success!')
+          }
+
+          setIsSubmitting(false)
+        }, 1000)
+      }
+    },
+    [email, password]
+  )
 
   return (
     <LayoutExternal>
       <Container>
         <FormWrapper>
           <Title>Sign in to Eventio.</Title>
-          {error ? (
-            <ErrorMessage>{error}</ErrorMessage>
+
+          {submitError ? (
+            <ErrorMessage>{submitError}</ErrorMessage>
           ) : (
             <Description>Enter your details below.</Description>
           )}
-          <form onSubmit={onSubmit}>
-            <Input label="Email" type="email" name="email" error={error} />
+
+          <form onSubmit={login}>
+            <Input
+              label="Email"
+              type="email"
+              name="email"
+              value={email}
+              error={emailError}
+              onChange={(e) => {
+                setEmailError(null)
+                setEmail(e.target.value)
+              }}
```

Comment:
As we are validating only on submit (which sucks, by the way, but we'll fix in the future), the best possible UX we could figure here is to ensure we clean any validation errors when the user modifies the input.

Reply:
I can imagine some kind of "rebounce(500ms)" type of utility would be the best UX (users get informed even before they tab out), but I wonder why did you decide not to use onBlur property, isn't that a tiny bit better than validating on submit?

Reply:
I can imagine some kind of "debounce(500ms)" type of utility would be the best UX (users get informed even before they tab out), but I wonder why did you decide not to use onBlur property, isn't that a tiny bit better than validating on submit?

Reply:
You are correct. Only then you’ll also have to validate on submit anyway. And suddenly you are repeating logic all around, doing very hard to read and imperative code.

You can check #11 on a more complete form state handling logic :)